Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp

```diff
@@ -2697,25 +2700,35 @@ Value* B3IRGenerator::emitCatchImpl(CatchKind kind, ControlType& data, unsigned
         connectControlAtEntrypoint(indexInBuffer, pointer, controlData, expressionStack, data);
     }
 
-    PatchpointValue* result = m_currentBlock->appendNew<PatchpointValue>(m_proc, m_proc.addTuple({ pointerType(), pointerType() }), origin());
+    PatchpointValue* result = m_currentBlock->appendNew<PatchpointValue>(m_proc, m_proc.addTuple({ pointerType(), B3::Int64 }), origin());
     result->effects.exitsSideways = true;
     result->clobber(RegisterSet::macroScratchRegisters());
     RegisterSet clobberLate = RegisterSet::volatileRegistersForJSCall();
     clobberLate.add(GPRInfo::argumentGPR0);
+    clobberLate.add(GPRInfo::argumentGPR1);
     result->clobberLate(clobberLate);
     result->append(instanceValue(), ValueRep::SomeRegister);
     result->resultConstraints.append(ValueRep::reg(GPRInfo::returnValueGPR));
-    result->resultConstraints.append(ValueRep::reg(GPRInfo::returnValueGPR2));
-    result->setGenerator([=] (CCallHelpers& jit, const StackmapGenerationParams& params) {
+    result->resultConstraints.append(B3::ValueRep::SomeRegister);
+    result->setGenerator([=] (auto& jit, const StackmapGenerationParams& params) {
         AllowMacroScratchRegisterUsage allowScratch(jit);
+        // Returning one EncodedJSValue on the stack
+        constexpr int32_t resultSpace = WTF::roundUpToMultipleOf<stackAlignmentBytes()>(static_cast<int32_t>(sizeof(EncodedJSValue)));
+        jit.subPtr(CCallHelpers::TrustedImm32(resultSpace), MacroAssembler::stackPointerRegister);
         jit.move(params[2].gpr(), GPRInfo::argumentGPR0);
+        jit.move(MacroAssembler::stackPointerRegister, GPRInfo::argumentGPR1);
         CCallHelpers::Call call = jit.call(OperationPtrTag);
         jit.addLinkTask([call] (LinkBuffer& linkBuffer) {
             linkBuffer.link(call, FunctionPtr<OperationPtrTag>(operationWasmRetrieveAndClearExceptionIfCatchable));
         });
+        if constexpr (is64Bit())
+            jit.loadValue(CCallHelpers::Address(MacroAssembler::stackPointerRegister), JSValueRegs(params[1].gpr()));
```

Comment:
So we just have an extra load on 64-bit now? I wonder if there is a better way

Reply:
I had previously figured this would be cold enough to overlook but you're right that it's worth fixing. Will have a fix ready soon.